Chapter two, first draft for review. (#10)
riscv-non-isa/riscv-security-model: specification/src/chapter2.adoc

* Chained +
An adversary is able to chain together multiple direct and indirect attacks to achieve a goal. For example, use a software interface exploit to affect a call stack, and use that to take redirect the control flow of a victim.

This specification is primarily concerned with ISA level mitigations against logical attacks.

Physical or remote attacks in general need to be addressed at system, protocol or governance level, and may require additional non-ISA mitigations. However, some ISA level mitigations can also help provide some mitigation against physical or remote attacks and this is indicated in the tables below.

The required level of protection can vary depending on use case. For example, a HW RoT may have stronger requirements on physical resistance than other parts of an SoC.

Finally, this specification does not attempt to rate attacks by severity, or by adversary skill level. Ratings tend to depend on use case specific threat models and requirements. 

==== Logical

[width=100%]
[%header, cols="5,5,5,10,15,10"]
|===
| ID#
| Attack   
| Type  
| Description

> | Risc-V mitigations 

Reviewer: ```suggestion
| Current RISC-V mitigations 
```

> | Planned mitigations

Reviewer: ```suggestion
| Planned RISC-V mitigations
```

| CAT_NNN
| Unrestricted access 
| Direct +
Logical
| Direct access to unauthroized resources in normal operation.

> a| * Risc-V privilege levels

Reviewer: ```suggestion
a| * RISC-V (V)U/(V)S/S(H)/M privilege levels
```

Author: I would prefer to just keep "Risc-V privilege levels". The full definition of those is later, I don't think we need to keep enumerating that everywhere.

> * Risc-V isolation

Reviewer: ```suggestion
* RISC-V supervisor domains (MTT)
```

Author: This is a catch-all for any Risc-V isolation mechanism. These are enumerated later. It's not just supervisor domains.

> * Risc-V virtualization

Reviewer: ```suggestion
* RISC-V hardware virtualization (H extension)
```

| 

| CAT_NNN
| Transient execution attacks
| Chained +
Logical
| Attacks on speculative execution implementations. 

> | Most of these attacks except Spectre v1 are specific to particular micro architectures, and Risc-V systems are expected not to be vulnerable. +

Reviewer: ```suggestion
| Known (documented) attacks except Spectre v1 are specific to particular micro-architectures, and RISC-V systems are not expected to be vulnerable to those. This is an evolving area of research. +
```

For example: +
https://meltdownattack.com/[Spectre and meltdown papers] +
 https://www.intel.com/content/www/us/en/developer/topic-technology/software-security-guidance/processors-affected-consolidated-product-cpu-model.html[Intel security guidance] +
https://developer.arm.com/documentation/#cf-navigationhierarchiesproducts=Arm%20Security%20Center,Speculative%20Processor%20Vulnerability[Arm speculative vulnerability]
| Fence.t could mitigate against Spectre v1.

| CAT_NNN
| Interface abuse
| Chained +
Logical
| Abusing interfaces across privilege or isolation boundaries, for example to elevate privilege or to gain unauthorized access to resources.

> a| * Risc V privilege levels

Reviewer: ```suggestion
a| * RISC-V (V)U/(V)S/S(H)/M privilege levels
```

Author: I would prefer to just keep "Risc-V privilege levels". The full definition of those is later, I don't think we need to keep enumerating that everywhere.

> * Risc-V isolation

Reviewer: ```suggestion
* RISC-V Supervisor domains (MTT)
```

Author: This is a catch-all for any Risc-V isolation mechanism. These are enumerated later. It's not just supervisor domains, and not just MTT.

Do you agree?

| High assurance cryptography

| CAT_NNN
| Event counting  
| Direct +
Logical
| For example, timing processes across privilege or isolation boundaries to derive information about confidential assets.
a| * Data-independent timing instructions
* Performance counters restricted by privilege and isolation boundaries (sscofpmf, smcntrpmf)
|

| CAT_NNN
| Redirect control flow
| Chained +
Logical
| Unauthorized manipulation of call stacks and jump targets to redirect a control flow to code controlled by an attacker. 

> a| * Shadow stacks

Reviewer: ```suggestion
a| * Shadow stacks (Zicfiss)
```

> * Landing pads

Reviewer: ```suggestion
* Landing pads (Zicfilp)
```

|

|===

==== Physical and remote

[width=100%]
[%header, cols="5,10,10,15,15"]
|===
| ID#
| Attack     
| Type 
| Description

> | Risc V recommendations 

Reviewer: ```suggestion
| RISC-V recommendations 
```

| CAT_NNN
| Analysis of physical leakage
| Direct or indirect +
Physical or remote
| For example, observing radiation, power line patterns, or temperature.  
a| * Implement robust power management and radiation control

> * Support data-independent timing instructions 

Reviewer: ```suggestion
* Data Independent Execution Latency (Zkt, Zvkt)
```

* Supervisor domain ID, privilege level, or MTT attributes, could be used to derive memory encryption contexts at domain or workload granularity
* Provide a degree of tamper resistance

| CAT_NNN
| Boot attacks
| Chained +
Logical or physical
a| * Glitching to bypass secure boot
* Retrieving residual confidential memory after a system reset
a| * Implement robust power management
* Implement cryptographic memory protection with at least boot freshness

| CAT_NNN
| Subverting supply chains
| Remote
| Infiltration or collusion to subvert security provisioning chains, software supply chains and signing processes, hardware supply chains, attestation processes
| Deploy appropriate governance, accreditation, and certification processes for an ecosystem.

|===

=== Ecosystem security objectives

Ecosystem security objectives identify a set of common features and mechanisms that can be used to enforce and establish trust in an ecosystem. 

These features are defined here at a functional level only. Technical requirements are typically use case specific and defined by external certification programmes. 

> In some cases Risc-V non-ISA specifications can provide guidance or protocols. This is discussed more in use case examples later in this specification.

Reviewer: ```suggestion
In some cases RISC-V non-ISA specifications can provide guidance or protocols. This is discussed more in use case examples later in this specification.
```

|===

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Generic security lifecycle"]
image::img_ch2_security-lifecycle.png[]

A security lifecycle reflects the trustworthiness of a system during its lifetime and reflects the lifecycle state of hardware provisioned assets. 

It can be extended as indicated below to cover additional security provisioning steps such as device onboarding, device activiation, user management, and RMA processes. These are use case or ecosystem specific and out of scope of this specification.

For the purpose of this specification, a minimum security lifecycle includes at least the following states:

* Manufacture - The system may not yet be fully locked down and has no hardware provisioned assets
* Security provisioning - The process of provisioning hardware provisioned assets +
Depending on ecosystem requirement, security provisioning could be performed in multiple stages through a supply chain and may require additional sub-states. These types of application specific extensions are out of scope of this specification.
* Secured - the system is fully locked down and has all its hardware provisioned assets +

> Additional application specific provisioning stages can take place in this state - for example network onboarding and device activation, or user identity management. This is out of scope of this specification.

Reviewer: ```suggestion
Additional application specific provisioning stages can take place in this state - for example network onboarding and device activation, TSS/App/Device attestation or user identity management. This is out of scope of this specification.
```

| CAT_NNN
| A layered attestation MUST be signed by a secure identity cryptographically bound (for example, hash locked) to a fresh security platform attestation

|===

NOTE: Care needs to be taken in attestation interface design. For example, software interfaces should only support either direct attestation or layered attestation workflows, never both, to prevent impersonation.

==== Authorized software 

Running unauthorized software can compromise the security state of the system. Two complementary processes can be used to authorize software:

* Measuring 
* Verification

A measurement is a cryptographic fingerprint of the launch state of a software component, such as a running hash of its memory contents and launch state.

Verification compares an actual measurement to an expected measurement from a signed authorization by a trusted signer. Verification requires a provisioned or securely discovered list of trusted signer(s). Depending on use case, a signed authorization can be part of a software image. Or it can be delivered or discovered separately as part of an authorization protocol.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | All isolated software that has to be trusted by other software on the same system MUST be measured and verified.

Reviewer: should verification be optional (may) since some platforms may find measured boot with attestation sufficient?

Author: I have updated the "Authorized software" section. Please review.

|===

Immutable boot code forms part of the immutable hardware security platform identified by the hardware security platform identity and is the start of a local trust chain on a system. It is typically ROM code, or locked on-chip flash.

If a system implements a hardware root of trust, then the HW root of trust hosts the immutable boot code. Otherwise the immutable boot code is typically the first code that executes on an application processor immediately after reset. 

==== Secure updates

Over time, any non-immutable component may need updates to address vulnerabilities or functionality improvements. An update can concern software/firmware, microcode, or any other updatable element on a system.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| All software on a system which is not immutable MUST be updatable.
|===

Immutable software includes at least immutable boot code. Some trusted subsystems can also include immutable software to meet specific security certification requirements defined by a governing body. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | An update of a software component SHOULD restart all other dependant software components.

Reviewer: this will be limiting capabilities that can securely update a layered TCB component without restarting dependent software component. (it is expected in these cases, the dependent SW components explicitly opt-in to updates, and the attestation reports that opt-in apriori along with the creation security version and the current security version of the TCB.

Author: This is the catch-all. Live-updates are covered just below. Are you ok with that?

Author: I have updated the "Authorized software" section above to prepare the ground. Please review.

Will tweak the update section tomorrow.

Author: Pushed an update to the system update section as well. Please review.

|===

A software update changes the attested security state of the affected component, and can affect whether the system is still considered trustworthy or not by a reliant party. A restart forces determination of a fresh attested state following the update. 

Alternatively, support for live updates may be part of the already attested trust contract between the reliant party and the system, given suitable governance. In which case a restart may not be required.

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  

> | An update of trusted security services MUST only take effect following a system reset.

Reviewer: most systems now want to update without a full reset.

Author: In server land maybe :-) 

Ok, I will review this section and suggest an update. We are basically talking about either local boot rooted in immutable boot code, where at least the trusted security services must be measure and verified; or host attestation where system boot only is measured, but the the host must be attested before launching hosted services.

Author: I have updated the "Authorized software" section above to prepare the ground. Please review.

Will tweak the update section tomorrow.

Reviewer 2: Also in other lands like industrial control systems that run mission critical processes.

Author: Pushed an update to the system update section as well. Please review.

[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN  
| Updates MUST be monotonic

| CAT_NNN
| Updates SHOULD be robust against update failures
|===

As a security principle and good practice it should not be possible to roll back to previous versions once an update has been locally accepted, as earlier versions may be carrying known vulnerabilities. For example using derived anti-rollback counters (counter tree) rooted in a hardware monotonic counter.

A system can still support recovery mechanisms, with suitable governance, in the case of update failures. 

[width=100%]
[%header, cols="5,20"]
|===
| ID#     
| Requirement

| CAT_NNN
| Updates and update messages MUST only be received from trusted sources.

|===

==== Isolation
Complex systems include software components from different supply chains, and complex integration chains with different roles and actors. These supply chains and integration actors often share mutual distrust:

* Developed, certified, deployed and attested independently
* Protected from errors in, or abuse from, other components
* Protected from debugging of other components
* Contain assets which should not be available to other components

> Use cases later in this specification provide examples of Risc-V isolation models.

Reviewer: ```suggestion
Use cases later in this specification provide examples of RISC-V isolation models.
```
